File: ports/1.21.1-full/src/main/java/mods/eln/ghost/GhostBlock.java
package mods.eln.ghost;

import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.level.block.RenderShape;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.phys.AABB;
import net.minecraft.core.BlockPos;
import net.minecraft.util.math.RayTraceResult;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;
import mods.eln.Eln;
import mods.eln.misc.Coordinate;
import mods.eln.misc.Direction;
import mods.eln.node.transparent.TransparentNodeEntity;
import net.minecraft.world.level.block.Block;
import net.minecraft.block.material.Material;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.phys.Vec3;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.Level;

import javax.annotation.Nullable;
import java.util.List;
import java.util.Random;

import net.minecraft.world.level.block.state.properties.IntegerProperty;
import net.minecraft.world.level.block.state.StateDefinition;

public class GhostBlock extends Block {

    public static final IntegerProperty META = IntegerProperty.create("meta", 0, 15);

    public static final int tCube = 0;
    public static final int tFloor = 1;
    public static final int tLadder = 2;

    public GhostBlock() {
        super(Material.GLASS);
        setDefaultState(blockState.getBaseState().withProperty(META, 0));
    }

    @Override
    protected StateDefinition createBlockState() {
        return new StateDefinition(this, META);
    }

    @Override
    public BlockState getStateFromMeta(int meta) {
        return getDefaultState().withProperty(META, meta);
    }

    @Override
    public int getMetaFromState(BlockState state) {
        return state.getValue(META);
    }

    @Nullable
    @Override
    public Item getItemDropped(BlockState state, Random rand, int fortune) {
        return null;
    }

    // TODO(1.10): Needs to be done by block states.
//    @Override
//    public void addCollisionBoxToList(IBlockState state, World world, BlockPos pos, AxisAlignedBB entityBox, List<AxisAlignedBB> collidingBoxes, @Nullable Entity entityIn) {
//        int meta = world.getBlockMetadata(x, y, z);
//
//        switch (meta) {
//            case tFloor:
//                AxisAlignedBB axisalignedbb1 = AxisAlignedBB.getBoundingBox((double) x, (double) y, (double) z, (double) x + 1, (double) y + 0.0625, (double) z + 1);
//                if (axisalignedbb1 != null && par5AxisAlignedBB.intersects(axisalignedbb1)) {
//                    list.add(axisalignedbb1);
//                }
//                break;
//            case tLadder:
//
//                break;
//            default:
//                GhostElement element = getElement(world, x, y, z);
//                Coordinate coord = element == null ? null : element.observatorCoordinate;
//                TileEntity te = coord == null ? null : coord.getTileEntity();
//                if (te != null && te instanceof TransparentNodeEntity) {
//                    ((TransparentNodeEntity) te).addCollisionBoxesToList(par5AxisAlignedBB, list, element.elementCoordinate);
//                } else {
//                    super.addCollisionBoxesToList(world, x, y, z, par5AxisAlignedBB, list, entity);
//                }
//                break;
//        }
//    }

//    @Override
//    @SideOnly(Side.CLIENT)
//    public AxisAlignedBB getSelectedBoundingBoxFromPool(World w, int x, int y, int z) {
//        int meta = w.getBlockMetadata(x, y, z);
//
//        switch (meta) {
//            case tFloor:
//                return AxisAlignedBB.getBoundingBox((double) x, (double) y, (double) z, (double) x + 1, (double) y + 0.0625, (double) z + 1);
//            case tLadder:
//                return AxisAlignedBB.getBoundingBox((double) x, (double) y, (double) z, (double) x + 0, (double) y + 0.0, (double) z + 0);
//            default:
//                return super.getSelectedBoundingBoxFromPool(w, x, y, z);
//        }
//    }
//
//    @Override
//    public MovingObjectPosition collisionRayTrace(World world, int x, int y, int z, Vec3d startVec, Vec3d endVec) {
//        int meta = world.getBlockMetadata(x, y, z);
//
//        switch (meta) {
//            case tFloor:
//                this.maxY = 0.0625;
//                break;
//            case tLadder:
//                this.maxX = 0.01;
//                this.maxY = 0.01;
//                this.maxZ = 0.01;
//                break;
//            default:
//                break;
//        }
//
//        MovingObjectPosition m = super.collisionRayTrace(world, x, y, z, startVec, endVec);
//
//        switch (meta) {
//            case tFloor:
//                this.maxY = 1;
//                break;
//            case tLadder:
//                this.maxX = 1;
//                this.maxY = 1;
//                this.maxZ = 1;
//                break;
//            default:
//                break;
//        }
//
//        return m;
//    }
//
//    @Override
//    public boolean isLadder(IBlockAccess world, int x, int y, int z, EntityLivingBase entity) {
//        return world.getBlockMetadata(x, y, z) == tLadder;
//    }


    @Override
    public int getLightOpacity(BlockState state) {
        return 0;
    }

    @Override
    public boolean isTranslucent(BlockState state) {
        return true;
    }

    @Override
    public boolean isOpaqueCube(BlockState state) {
        return false;
    }

    @Override
    public boolean isFullCube(BlockState state) {
        return false;
    }

    @Override
    public RenderShape getRenderType(BlockState state) {
        return RenderShape.INVISIBLE;
    }

    @Override
    public ItemStack getPickBlock(BlockState state, RayTraceResult target, Level world, BlockPos pos, Player player) {
        return ItemStack.EMPTY;
    }

    @Override
    public boolean isTopSolid(BlockState state) {
        return false;
    }

    @Override
    public boolean isSideSolid(BlockState base_state, BlockGetter world, BlockPos pos, net.minecraft.core.Direction side) {
        return false;
    }

    @Override
    public void breakBlock(Level world, BlockPos pos, BlockState state) {
        if (!world.isRemote) {
            GhostElement element = getElement(world, pos);
            if (element != null) {
                element.breakBlock();
            }
        }
        super.breakBlock(world, pos, state);
    }

    @Override
    public boolean onBlockActivated(Level world, BlockPos pos, BlockState state, Player player, InteractionHand hand, net.minecraft.core.Direction facing, float hitX, float hitY, float hitZ) {
        if (!world.isRemote) {
            GhostElement element = getElement(world, pos);
            if (element != null)
                return element.onBlockActivated(player, Direction.fromFacing(facing), hitX, hitY, hitZ);
        }
        return true;
    }

    private GhostElement getElement(Level world, BlockPos pos) {
        return Eln.ghostManager.getGhost(new Coordinate(pos, world));
    }

    @Override
    public float getBlockHardness(BlockState blockState, Level worldIn, BlockPos pos) {
        return 0.5f;
    }

    public String getNodeUuid() {
        return "g";
    }
}

File: src/main/kotlin/mods/eln/ghost/GhostBlock.kt
@file:Suppress("NAME_SHADOWING")
package mods.eln.ghost

import mods.eln.Eln
import mods.eln.misc.Coordinate
import mods.eln.misc.Direction.Companion.fromFacing
import mods.eln.node.transparent.TransparentNodeEntity
import net.minecraft.world.level.block.Block
import net.minecraft.block.material.Material
import net.minecraft.world.level.block.state.properties.IntegerProperty
import net.minecraft.block.state.BlockFaceShape
import net.minecraft.world.level.block.state.StateDefinition
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.block.RenderShape
import net.minecraft.core.Direction as EnumFacing
import net.minecraft.world.InteractionHand
import net.minecraft.world.phys.AABB
import net.minecraft.core.BlockPos
import net.minecraft.world.phys.HitResult
import net.minecraft.world.phys.Vec3
import net.minecraft.world.level.BlockGetter
import net.minecraft.world.level.Level
import java.util.Random

/**
 * The invisible placeholder a multiblock machine puts in the cells it occupies.
 *
 * Its shape used to be its metadata value ([tCube], [tFloor], [tLadder]); on 1.12 that is a
 * blockstate property. The shape also used to be applied by mutating the block's own
 * `maxX`/`maxY`/`maxZ` fields around a `super.collisionRayTrace` call - those fields no longer
 * exist, because a block's bounds are now derived from its state, so each shape returns its own
 * bounding box instead.
 */
class GhostBlock : Block(Material.IRON) {

    init {
        defaultState = blockState.baseState.withProperty(SHAPE, tCube)
    }

    override fun createBlockState(): StateDefinition = StateDefinition(this, SHAPE)

    override fun getStateFromMeta(meta: Int): BlockState =
        defaultState.withProperty(SHAPE, meta.coerceIn(tCube, tLadder))

    override fun getMetaFromState(state: BlockState): Int = state.getValue(SHAPE)

    private fun shapeOf(state: BlockState) = state.getValue(SHAPE)

    override fun getBoundingBox(state: BlockState, world: BlockGetter, pos: BlockPos): AABB =
        when (shapeOf(state)) {
            tFloor -> FLOOR_BOX
            tLadder -> LADDER_BOX
            else -> FULL_BLOCK_AABB
        }

    override fun getItemDropped(state: BlockState, random: Random, fortune: Int): Item? = null

    override fun addCollisionBoxToList(
        state: BlockState, world: Level, pos: BlockPos, entityBox: AABB,
        collidingBoxes: MutableList<AABB>, entity: Entity?, isActualState: Boolean
    ) {
        when (shapeOf(state)) {
            tFloor -> addCollisionBoxToList(pos, entityBox, collidingBoxes, FLOOR_BOX)
            tLadder -> Unit  // a ladder ghost is passable
            else -> {
                val element = getElement(world, pos.x, pos.y, pos.z)
                val te = element?.observatorCoordonate?.tileEntity
                if (te is TransparentNodeEntity) {
                    @Suppress("UNCHECKED_CAST")
                    te.addCollisionBoxesToList(
                        entityBox, collidingBoxes as MutableList<AABB?>, element.elementCoordinate
                    )
                } else {
                    super.addCollisionBoxToList(state, world, pos, entityBox, collidingBoxes, entity, isActualState)
                }
            }
        }
    }

    override fun getSelectedBoundingBox(state: BlockState, world: Level, pos: BlockPos): AABB =
        when (shapeOf(state)) {
            tFloor -> FLOOR_BOX.offset(pos)
            tLadder -> LADDER_BOX.offset(pos)
            else -> super.getSelectedBoundingBox(state, world, pos)
        }

    override fun collisionRayTrace(
        blockState: BlockState, world: Level, pos: BlockPos, startVec: Vec3, endVec: Vec3
    ): HitResult? = rayTrace(pos, startVec, endVec, getBoundingBox(blockState, world, pos))

    override fun isLadder(state: BlockState, world: BlockGetter, pos: BlockPos, entity: LivingEntity?): Boolean =
        shapeOf(state) == tLadder

    override fun isOpaqueCube(state: BlockState): Boolean = false

    override fun isFullCube(state: BlockState): Boolean = false

    override fun getRenderType(state: BlockState): RenderShape = RenderShape.INVISIBLE

    override fun getPickBlock(
        state: BlockState, target: HitResult, world: Level, pos: BlockPos, player: Player
    ): ItemStack = ItemStack.EMPTY

    override fun getBlockFaceShape(
        world: BlockGetter, state: BlockState, pos: BlockPos, face: EnumFacing
    ): BlockFaceShape = BlockFaceShape.UNDEFINED

    override fun breakBlock(world: Level, pos: BlockPos, state: BlockState) {
        if (!world.isClientSide) {
            getElement(world, pos.x, pos.y, pos.z)?.breakBlock()
        }
        super.breakBlock(world, pos, state)
    }

    override fun onBlockActivated(
        world: Level, pos: BlockPos, state: BlockState, player: Player,
        hand: InteractionHand, side: EnumFacing, vx: Float, vy: Float, vz: Float
    ): Boolean {
        if (!world.isClientSide) {
            val element = getElement(world, pos.x, pos.y, pos.z)
            if (element != null) return element.onBlockActivated(player, fromFacing(side), vx, vy, vz)
        }
        return true
    }

    fun getElement(world: Level?, x: Int, y: Int, z: Int): GhostElement? {
        return Eln.ghostManager.getGhost(Coordinate(x, y, z, world!!))
    }

    override fun getBlockHardness(blockState: BlockState, world: Level, pos: BlockPos): Float = 0.5f

    val nodeUuid: String
        get() = "g"

    companion object {
        const val tCube = 0
        const val tFloor = 1
        const val tLadder = 2

        @JvmField
        val SHAPE: IntegerProperty = IntegerProperty.create("shape", tCube, tLadder)

        private val FLOOR_BOX = AABB(0.0, 0.0, 0.0, 1.0, 0.0625, 1.0)
        private val LADDER_BOX = AABB(0.0, 0.0, 0.0, 0.0, 0.0, 0.0)
    }
}
